### DISCUSSION TITLE
Narrowing sources' falsyness

### DISCUSSION THREAD

**Discussion by XiNiHa** [TanStack/query#6464] [Ideas]
When writing dependent queries, it's common to expect the query function to depend on the value passed to the `enabled` option and its truthyness. If `useQuery` can ensure the behavior with some type narrowing, it would make all falsyness checks in the query function redundant. For example:

```tsx
const { data: user }  = useQuery({
  ...opts,
  queryFn: fetchUser,
})
const { data: organization } = useQuery({
  // the name of the option wouldn't make sense anymore and would need a new one if adopted, like `source`
  enabled: user,
  // ctx.enabled is typed as `Exclude<typeof user, Falsy>`, ensuring that the value is truthy and need no null check
  queryFn: ({ enabled: user }) => fetchOrganization(user.organizationId)
})
```

This design is already adopted in [SolidJS's `createResource()` (the overload that uses a source signal)](https://www.solidjs.com/docs/latest/api#createresource) and works pretty conveniently.

**Comment by TkDodo** [comment]
this looks interesting
